operators/src/engine/execution_context.rs

```diff
@@ -1,46 +1,163 @@
-use geoengine_datatypes::{
-    primitives::Coordinate2D,
-    raster::{GridShape2D, TilingSpecification},
-};
-
 use crate::concurrency::{ThreadPool, ThreadPoolContext};
+use crate::engine::{QueryRectangle, VectorResultDescriptor};
+use crate::error::Error;
+use crate::mock::MockDataSetDataSourceLoadingInfo;
+use crate::source::OgrSourceDataset;
+use crate::util::Result;
+use geoengine_datatypes::dataset::DataSetId;
+use geoengine_datatypes::raster::GridShape;
+use geoengine_datatypes::raster::TilingSpecification;
+use std::any::Any;
+use std::collections::HashMap;
+use std::fmt::Debug;
 use std::path::PathBuf;
 
-/// A context that provides certain utility access during operator execution
-/// TODO: maybe this is rather a trait
-#[derive(Debug, Clone)]
-pub struct ExecutionContext<'pool> {
-    pub raster_data_root: PathBuf,
-    pub thread_pool: ThreadPoolContext<'pool>,
-    pub tiling_specification: TilingSpecification,
+/// A context that provides certain utility access during operator initialization
+pub trait ExecutionContext:
+    Send
+    + Sync
+    + LoadingInfoProvider<MockDataSetDataSourceLoadingInfo, VectorResultDescriptor>
+    + LoadingInfoProvider<OgrSourceDataset, VectorResultDescriptor>
+{
+    fn thread_pool(&self) -> ThreadPoolContext;
```

Comment:
```suggestion
    fn thread_pool(&self) -> &ThreadPoolContext;
```
I guess it's sufficient to hand out a reference to it. You can clone a context (currently) but I guess you don't want to all the time.

Reply:
but who owns the contexts if I only hand out a reference? There can be multiple in each execution context or not?

Reply:
The goal is to have a single context for each execution unit, e.g., computation of a workflow.
Then, each operator can make use of this joint context and schedule work. This allows us to later on schedule differently. For now, we know that all work items share the same "group id" in the ThreadPoolContext.

Reply:
Do we really want one group per ExecutionContext? Maybe we want one per user or user group.  Or one for anonymous user and one for registered users. The execution context could abstract that.

If we decide to hand out only references, then the ExecutionContext has to own the ThreadPoolContext which introduces a lifetime to the trait. Also for the Mock-implementation, the struct has to own the ThreadPool AND the ThreadPoolContext (which references the ThreadPool) in order to allow for convenient Default-constructor. As you know referencing a field of the struct itself is not possible. I could turn the ThreadPoolContext into a trait and have an OwningThreadPoolContext which is a single ThreadPoolContext that owns the thrad pool, which might be sufficient in tests. But then the ExecutionContext has to hand out a &dyn ThreadPoolContext a fat pointer you two don't like.

Shall I do it or do you have other ideas to solve this?

Reply:
> Do we really want one group per ExecutionContext? Maybe we want one per user or user group. Or one for anonymous user and one for registered users. The execution context could abstract that.

It is not clear yet, but that doesn't change that we don't necessarily want to create a thread context on each call.

> If we decide to hand out only references, then the ExecutionContext has to own the ThreadPoolContext which introduces a lifetime to the trait.

Actually, the implementation would need one, but not the trait.

> Also for the Mock-implementation, the struct has to own the ThreadPool AND the ThreadPoolContext (which references the ThreadPool) in order to allow for convenient Default-constructor. As you know referencing a field of the struct itself is not possible.

Yeah, that was the reason for the implementation before you changed it in this PR.

> I could turn the ThreadPoolContext into a trait and have an OwningThreadPoolContext which is a single ThreadPoolContext that owns the thrad pool, which might be sufficient in tests. But then the ExecutionContext has to hand out a &dyn ThreadPoolContext a fat pointer you two don't like.

That is not the only solution. We could also think about…
 * … having a creator like before,
 * … introducing an Arc of the pool inside the context to get rid of the lifetime
 * …

> Shall I do it or do you have other ideas to solve this?

If it's not easy to find a good solution, we should discuss what we want or know for now and then do it.

Reply:
The ThreadPoolContext is a ThreadPool-reference and a group id. I agree that multiple users of the thread-pool in the scope of a request may share the same group id but
- I'm not sure that it's always the same group id
- the overhead of having multiple individual contexts with the same group id is really really negligible
- owning the context allows for a cleaner design because it takes the ExecutionContext out of the picture. Operators want to keep the threadPoolContext longer than the execution context, which is only needed in the `initialize` method

I vote for keeping the trait like it is or creating an extra task for this.